File: src/app/src/main/java/com/huawei/podcast/java/utils/AudioTrackThread.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2019-2019. All rights reserved.
 */

package com.huawei.podcast.java.utils;

import android.media.AudioFormat;
import android.media.AudioManager;
import android.media.AudioTrack;
import android.util.Log;

/**
 * 	function description
 */
public class AudioTrackThread extends Thread {

    private static final int DEFAULT_INT_TYPE = Constants.ZERO;

    private static final int SAMPLE_RATE = Constants.SAMPLE_RATE;

    private boolean mIsRunning = true;

    private AudioTrack mAudioTrack = null;

    @Override
    public void run() {
        setPriority(Thread.MAX_PRIORITY);
        int buffsize =
                AudioTrack.getMinBufferSize(SAMPLE_RATE, AudioFormat.CHANNEL_IN_STEREO, AudioFormat.ENCODING_PCM_16BIT);
        try {
            mAudioTrack =
                    new AudioTrack(
                            AudioManager.STREAM_MUSIC,
                            SAMPLE_RATE,
                            AudioFormat.CHANNEL_IN_STEREO,
                            AudioFormat.ENCODING_PCM_16BIT,
                            buffsize,
                            AudioTrack.MODE_STREAM);
        } catch (IllegalThreadStateException e) {
        }
        if (mAudioTrack == null) {
            return;
        }
        audioTrackWrite(buffsize);
        mAudioTrack.release();
    }

    private void audioTrackWrite(int buffsize) {
        short[] samples = new short[buffsize];
        if (mAudioTrack.getState() != AudioTrack.STATE_INITIALIZED) {
        } else {
            mAudioTrack.play();
            try {
                while (mIsRunning) {
                    mAudioTrack.write(samples, DEFAULT_INT_TYPE, buffsize);
                }
            } catch (IllegalThreadStateException e) {
            }
            mAudioTrack.stop();
        }
    }

    /**
     * thread destroy
     */
    public void destroy() {
        if (mAudioTrack != null) {
            mAudioTrack.stop();
        }
        mIsRunning = false;
    }
}

File: src/app/src/main/java/com/huawei/podcast/kotlin/utils/AudioTrackThread.kt
/**
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.huawei.podcast.kotlin.utils

import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioTrack

/**
 * function description
 */
class AudioTrackThread : Thread() {
    private var mIsRunning = true
    private var mAudioTrack: AudioTrack? = null

    override fun run() {
        priority = MAX_PRIORITY
        val buffSize = AudioTrack.getMinBufferSize(SAMPLE_RATE, AudioFormat.CHANNEL_IN_STEREO, AudioFormat.ENCODING_PCM_16BIT)
        try {
            mAudioTrack = AudioTrack(
                AudioManager.STREAM_MUSIC, SAMPLE_RATE, AudioFormat.CHANNEL_IN_STEREO,
                AudioFormat.ENCODING_PCM_16BIT, buffSize, AudioTrack.MODE_STREAM
            )
        } catch (e: IllegalThreadStateException) {
        }
        if (mAudioTrack == null) {
            return
        }
        audioTrackWrite(buffSize)
        mAudioTrack?.release()
    }

    private fun audioTrackWrite(buffSize: Int) {
        val samples = ShortArray(buffSize)
        if (mAudioTrack?.state != AudioTrack.STATE_INITIALIZED) {
        } else {
            mAudioTrack?.play()
            try {
                while (mIsRunning) {
                    mAudioTrack?.write(samples, DEFAULT_INT_TYPE, buffSize)
                }
            } catch (e: IllegalThreadStateException) {
            }
            mAudioTrack?.stop()
        }
    }

    /**
     * thread destroy
     */
    override fun destroy() {
        if (mAudioTrack != null) {
            mAudioTrack?.stop()
        }
        mIsRunning = false
    }

    companion object {
        private const val DEFAULT_INT_TYPE = 0
        private const val SAMPLE_RATE = 44100
    }
}
